Pre-compile client-side EJS templates

seanmonstar opened the pull request:
All EJS templates that are used client-side (mostly in the dialog) are now compiled into functions _before_ being sent to the client. In dev, this file is dynamically generated each time, and for production, the compress script continues to take care of everything.

That means no more EJS client-side lib. Also, removed an instance of `_.template`, since that also `eval`s code, and that's no fun.

seanmonstar committed:
updating ejs to needed version for client compiling

seanmonstar committed:
/common/js/templates.js gets built dynamically in dev

The list of templates gets pushed onto a single file, with the EJS pre-compiled
into functions. Then, in the Renderer, the functions only need to be invoked,
meaning we no longer need the EJS client lib, and don't eval code.

seanmonstar committed:
updating templates, since <%= escapes content by default now

seanmonstar committed:
adding /test/mocks/site-templates.js route

this returns compiled site templates, similar to the dialog templates,
which are needed by the test suite to test the site works!

seanmonstar committed:
remove last instances of EJS on client-side

manage acct page uses a template, but has no
need for EJS to parse it, since it's plain HTML.

seanmonstar committed:
remove _.template usage

_.template uses new Function, which means evalling code
evalling code is bad. so now we don't do it.

seanmonstar committed:
compileDebug should be false is minified

seanmonstar committed:
update static resources test since ejs is gone

seanmonstar committed:
resolving merge conflicts

shane-tomlinson commented on package.json:
```diff
@@ -13,7 +13,7 @@
         "connect-cookie-session": "0.0.2",
         "connect-logger-statsd": "0.0.1",
-        "ejs": "0.4.3",
+        "ejs": "git://github.com/visionmedia/ejs.git#3c3f592d2f261bd2239653839c85b8bc90833c86",
```
It does not look like a new npm version of ejs has been published yet, npm version shows the last version was cut on 2012-08-16. Once a new version is published, let's change this sha to a version.

shane-tomlinson replied:
@seanmonstar - https://github.com/visionmedia/ejs/pull/66#issuecomment-8536481

0.8.3 is now published.

shane-tomlinson commented on lib/static_resources.js:
```diff
@@ -20,7 +20,6 @@ var common_js = [
   '/common/js/lib/winchan.js',
   '/common/js/lib/underscore.js',
   '/common/js/lib/bidbundle.js',
-  '/common/js/lib/ejs.js',
```
This is awesome. Can you also remove it from resources/test.ejs? One of the many non-DRY failures in action is that many includes are declared in both resources/test.ejs and lib/static_resources.js

shane-tomlinson commented on resources/static/common/js/renderer.js:
```diff
@@ -7,34 +7,16 @@ BrowserID.Renderer = (function() {
-    var html = template.render(vars);
-    return html;
+    // arguments are: locals, filters (which cant be used client-side), escapeFn
+    return templateFn.call(null, vars);
```
Is the escapeFn ever used in templateFn? Does it need to be passed in?

seanmonstar replied:
EJS has a default escape function that gets included, and we have the ability to override it: https://github.com/visionmedia/ejs/blob/master/lib/utils.js#L16

shane-tomlinson commented on resources/static/dialog/views/confirm_email.ejs:
```diff
@@ -5,10 +5,10 @@
 
     <p>
-      <%= format(gettext('Click the link in the confirmation email. You\'ll then immediately be signed in to %s.'), ["<strong>" + siteName + "</strong>"]) %>
+      <%- format(gettext('Click the link in the confirmation email. You\'ll then immediately be signed in to %s.'), ["<strong>" + escape(siteName) + "</strong>"]) %>
```
siteName should already be escaped by this point - https://github.com/mozilla/browserid/blob/5b1150b3d52c65c1e5e591e5a42474ec508ac68e/resources/static/dialog/js/modules/dialog.js#L201-203

This is going to cause a double-escaping of siteName.

emails are not escaped whenever they are loaded, so they _should_ be escaped when they are displayed, siteName will already have been escaped so I do not believe they should be when displayed.

@warner - can you weigh in here?

seanmonstar replied:
OK, if we are for sure escaping it earlier on the client-side, then that is fine.

This was just my knee-jerk dirty-in, dirty-out behavior.

shane-tomlinson commented on resources/static/pages/js/manage_account.js:
```diff
@@ -79,16 +79,19 @@ BrowserID.manageAccount = (function() {
-    _.templateSettings = {
-        interpolate : /\{\{(.+?)\}\}/g
-    };
+    function substitute(text, values, re) {
```
@seanmonstar - The original reason for not using an external template for templateManage and templateUser is because we did not include the external templates on the main site. All templates are now included everywhere to provide a common set of templates for error/wait messages and tooltips. 

Since a large change is already necessary here, now would be a good time to make a simplification and remove the manage_account specific hokiness. These templates could be moved to external files and normal rendering mechanism used to display templateManage and templateUser.

seanmonstar replied:
True, except that this is a `page`, and so it's templates are expected to render server-side, and are not included in the full Templates.js package. If it were, it would be more bytes for the dialog that are never used...

shane-tomlinson replied:
That's a good point, the reverse is true right now. On the main site all of the dialog templates are included, which bloats the main site slightly. I opened a new issue (#2481) for this.

We should figure out a good strategy to only serve up the needed templates for the context.

shane-tomlinson commented on lib/static/views.js:
```diff
@@ -230,6 +230,28 @@ exports.setup = function(app) {
 
+  // /common/js/templates.js is dynamically built each time
+  if (!config.get('use_minified_resources')) {
+    var generateTemplates = require('../../scripts/create_templates');
```
Looking at the overall directory layout and how requires are typically included, I feel like it would be preferable to put all of the create_templates code into a file in the /lib directory that can be included from both scripts/create_templates and here.

shane-tomlinson commented on lib/static/views.js:
```diff
@@ -230,6 +230,28 @@ exports.setup = function(app) {
+
+    var siteTemplatesPath = path.join(__dirname, "../../resources/views");
+    var siteTemplatesData;
+    app.get('/test/mocks/site-templates.js', function(req, res) {
```
The site/ link was used for unit testing so that we could dynamically render main site templates to the DOM between each test. With the dynamic creation of templates, this should no longer be necessary.

Can we use this as an opportunity to remove the site/ link from the resources/static/dialog/views directory?

shane-tomlinson commented on resources/views/test.ejs:
```diff
@@ -84,11 +84,13 @@
     <script src="mocks/mocks.js"></script>
     <script src="mocks/xhr.js"></script>
     <script src="mocks/templates.js"></script>
+    <script src="mocks/site-templates.js"></script>
```
++ awesome

shane-tomlinson commented on scripts/create_templates.js:
```diff
@@ -6,19 +6,31 @@
 
```
See comment above about moving a large portion of this into a file in the /lib directory that can be included from both views.js and here.

shane-tomlinson commented on tests/static-resource-test.js:
```diff
@@ -57,7 +57,7 @@ suite.addBatch({
       // Fragile - filename with :locale...
       // When fixing this test case... console.log(res[Object.keys(res)[0]]);
-      var localeIndex = 9;
+      var localeIndex = 8;
```
What does this do?

seanmonstar replied:
It's awesome. Let me tell you a story.

This is a variable that contains the index of 'i18n/:locale/...', that we have to count by hand after a `console.log`. Since I removed the EJS lib, there was one less resource in the list before the :locale resource. Hence the "fragile" comment above.

shane-tomlinson replied:
Oh wow. That is FANTASTIC. **&lt;face palm /&gt;**

seanmonstar committed:
review fixes

- dont escape siteName, we escape it right when we get it
- no longer need site views symlink
- remove ejs lib from /test

seanmonstar committed:
split up create_templates script into a lib file

seanmonstar committed:
update ejs@0.8.3

seanmonstar committed:
un-fragile the :locale static resources test

seanmonstar commented:
@shane-tomlinson fixed up those things you mentioned.

bonus points: fixed the locale static test by using indexOf.

shane-tomlinson commented:
@seanmonstar - The change to use indexOf is great. Testing the rest now.

shane-tomlinson commented on resources/static/pages/js/manage_account.js:
```diff
@@ -79,16 +79,19 @@ BrowserID.manageAccount = (function() {
-      var e = item.address,
-          identity = _.template(template, { email: e });
+      var e = item.address;
+      var id = e.replace('@', '_').replace('.', '_', 'g');
```
This is a smart change.

seanmonstar replied:
Though, I wondered: what about international emails? Or emails with other strange characters in them? We'll still have an illegal id.

warner replied:
I don't yet grok how these IDs are used or what the rules are, but if they only need to be unique and mapped to the original email addresses, could you replace them with sequential integers? That'd always be safe.

seanmonstar replied:
I don't fully know what the IDs are needed for, I just pulled the generation of the ID from out of the template code and into here.

shane-tomlinson commented on lib/static/views.js:
```diff
@@ -230,6 +230,28 @@ exports.setup = function(app) {
+    var dialogTemplatesPath = path.join(__dirname, '../../resources/static/dialog/views')
+    var dialogTemplatesData;
+    app.get('/common/js/templates.js', function(req, res) {
+      var str = templates.generate(dialogTemplatesPath);
```
Is the intention between L239 and 243 to only update dialogTemplatesData if templates.generate returns a value? Is it possible for templates.generate to return a value on one run but not a subsequent one?

Would this be the case if a template is valid the first time it is compiled, a developer edits it into an invalid state, and an attempt is made to recompile?

seanmonstar replied:
This case actually really only happens for the `create_templates` script, since it passes a `lastGen` timestamp based on `fs.statSync('build/templates.js').mtime`. If none of the individual template files have changed since then, it won't rebuild. Since it's not a long running process, it won't have the data cached in the closure, like it does when running the server.

shane-tomlinson commented on lib/templates.js:
```diff
@@ -0,0 +1,61 @@
+      var templateName = namePrefix + fileName.replace(/\.ejs/, '');
+      var templateText = fs.readFileSync(path.join(templatesDir, fileName), "utf8");
+
+      templates[templateName] = ejs.compile(templateText, {
```
Can EJS throw an exception or anything that would bring the server down in development mode?

seanmonstar replied:
I imagine it can. I can add a try/catch in the routes, with a log. I think the exception is worth showing in full when running `scripts/compress`, though.

seanmonstar replied:
Actually, after intentionally breaking a template and testing, it doesn't take down the server. The error gets logged normally. I imagine this is due to our `process.on('uncaughtException')` handler.

shane-tomlinson commented on lib/static/views.js:
```diff
@@ -230,6 +230,28 @@ exports.setup = function(app) {
 
+  // /common/js/templates.js is dynamically built each time
+  if (!config.get('use_minified_resources')) {
+    var templates = require('../templates');
```
I am wondering if there is a cleaner way to load the templates but still produce the desired results.

Instead of the extra housekeeping, it would be nice to be able to do something like:

```
res.send(templates.dialog())
```

Adding the extra functionality to templates.js feels like it would junk that abstraction up, so I am not sure what the best way to approach this is.

seanmonstar replied:
I could add `dialog` and `site` methods to the `templates` module. It would simply be moving the code from one file to another.

seanmonstar commented on resources/static/pages/js/manage_account.js:
```diff
@@ -79,16 +79,19 @@ BrowserID.manageAccount = (function() {
-          identity = _.template(template, { email: e });
+      var e = item.address;
+      var id = e.replace('@', '_').replace('.', '_', 'g');
+      var identity = substitute(template, { email: e, id: id });
```
I didn't even notice this: the email and id aren't escaped. :(

seanmonstar replied:
Though, I see they are pulled from localStorage. Are emails escaped before being placed there? Would my passed the email to `_.escape` be double escaping it?

warner replied:
um, is "substitute" a plain string interpolation? Yeah, let's fix that.

seanmonstar replied:
Yea, `substitute` is defined just a few lines above. I just want to be sure that I won't be double-escaping.

seanmonstar committed:
templateData is always returned if no lastGen is passed

seanmonstar committed:
templateData is always returned if no lastGen is passed

shane-tomlinson commented:
@seanmonstar - I am trying to figure out a way to easily do partial support for #2486. Can you check out this gist and see what you think:

https://gist.github.com/3736556

Commit:
Add partial support to client side templates.

Commit:
Add unit tests checking for partial support on the client.

shane-tomlinson commented:
@seanmonstar - instead of doing the partial support in gists, I added https://github.com/seanmonstar/browserid/pull/3

Commit:
Remove HTML comments to trim down the amount of stuff sent to the client.

Commit:
No longer use the email address as an element ID.

* Unit tests now check the displayed order of the addresses against the data model's sorted order.

seanmonstar committed:
Merge pull request #4 from shane-tomlinson/remove_html_comments

Remove HTML comments from compiled templates

seanmonstar committed:
Merge pull request #5 from shane-tomlinson/email_ids

No longer use the email address as an element ID.

seanmonstar committed:
Merge pull request #3 from shane-tomlinson/precompile-ejs

Add partial support to client side templates.

seanmonstar committed:
escape email address on manage account page

seanmonstar commented:
@shane-tomlinson this is loaded with a bunch of new awesomeness. Are we good to go?
